### PR TITLE
Per-map hash seeds

### PR DESCRIPTION
**Pull request by dgryski**
Created on top of https://github.com/tinygo-org/tinygo/pull/2687

### CODE REVIEW

**Comment by dgryski** [comment]
Draft because it's still failing tests.

**Comment by dgryski** [comment]
Bah, these commits ended up with too many changes.  I'm going to split out a few of them into a smaller commit.

**Comment by dgryski** [comment]
I don't understand this CI failure.

**Comment by dgryski** [comment]
Pulled out the hash function changes to see if that's what CI doesn't like.

**Comment by dgryski** [comment]
Huh, so indeed something about the new hash functions causes the CI issue.  I'll keep that out for now.

**Comment by deadprogram** [comment]
@dgryski can you now rebase this PR and resolve merge conflicts? Thanks!

**Comment by dgryski** [comment]
Rebased against `dev` now that map growth PR has landed.

**Comment by dgryski** [comment]
ping @aykevl

File: src/runtime/hashmap.go [hunk 1]
```diff
@@ -413,15 +418,15 @@ func hashmapBinaryGet(m *hashmap, key, value unsafe.Pointer, valueSize uintptr)
...
 	}
-	hash := hash32(key, uintptr(m.keySize))
+	hash := hash32(key, uintptr(m.keySize), m.seed)
 	return hashmapGet(m, key, value, valueSize, hash)
 }
 
 func hashmapBinaryDelete(m *hashmap, key unsafe.Pointer) {
 	if m == nil {
 		return
 	}
```
Comment (aykevl) [inline src/runtime/hashmap.go:428]: Now that all the `hashmap*Set`, `hashmap*Get`, and `hashmap*Delete` functions have nil checks themselves, I think the nil checks can be removed from the functions `hashmapSet`, `hashmapGet`, and `hashmapDelete`.

**Review by aykevl — COMMENTED** [review]
The code itself looks reasonable, with one nit below. I've also checked the binary size impact and it's actually quite minor. Of all the driver examples, only a few changed size (apparently maps aren't used much in driver tests). And for the regular tests, the biggest increases were in `testdata/testing.go`, `testdata/map.go`, `testdata/reflect.go`, which is either somewhat unusual Go code (map, reflect) or simply very large (testing).

Can you explain a bit more about why this is needed? Is it for the [hashmap DoS attack](https://lwn.net/Articles/474912/)? I'm generally hesitant with merging things that increase code size, unless there's a clear benefit.

**Comment by dgryski** [comment]
Yes, this is to protect against hash flooding attacks. While less important for microcontroller targets, the people running wasm web services will find this a welcome security fix. Incoming HTTP Headers, for example, are generally stored in maps and being user controlled makes them an easy avenue to exploit.

**Review by aykevl — APPROVED** [review]
Thank you for the explanation!

**Comment by deadprogram** [comment]
:tada: thank you @dgryski now merging!

**Pull request merged**